File: misc/BMP280/python/BMP280.py
import time
import smbus

# BMP280 address.
BMP280_I2C_ADDRESS       =    0x76          #SDO = 0

#Registers  value
BMP280_ID_Value          =    0x58
BMP280_RESET_VALUE       =    0xB6

# BMP280 Registers definition  
BMP280_TEMP_XLSB_REG     =    0xFC		  #Temperature XLSB Register 
BMP280_TEMP_LSB_REG      =    0xFB         #Temperature LSB Register  
BMP280_TEMP_MSB_REG      =    0xFA         #Temperature LSB Register   
BMP280_PRESS_XLSB_REG    =    0xF9		  #Pressure XLSB  Register 
BMP280_PRESS_LSB_REG     =    0xF8		  #Pressure LSB Register 
BMP280_PRESS_MSB_REG     =    0xF7		  #Pressure MSB Register 						
BMP280_CONFIG_REG        =    0xF5         #Configuration Register 
BMP280_CTRL_MEAS_REG     =    0xF4         #Ctrl Measure Register  
BMP280_STATUS_REG        =    0xF3         #Status Register 
BMP280_RESET_REG         =    0xE0		  #Softreset Register  
BMP280_ID_REG            =    0xD0         #Chip ID Register 

#calibration parameters   
BMP280_DIG_T1_LSB_REG    =    0x88  
BMP280_DIG_T1_MSB_REG    =    0x89  
BMP280_DIG_T2_LSB_REG    =    0x8A  
BMP280_DIG_T2_MSB_REG    =    0x8B  
BMP280_DIG_T3_LSB_REG    =    0x8C  
BMP280_DIG_T3_MSB_REG    =    0x8D  
BMP280_DIG_P1_LSB_REG    =    0x8E  
BMP280_DIG_P1_MSB_REG    =    0x8F  
BMP280_DIG_P2_LSB_REG    =    0x90  
BMP280_DIG_P2_MSB_REG    =    0x91  
BMP280_DIG_P3_LSB_REG    =    0x92  
BMP280_DIG_P3_MSB_REG    =    0x93  
BMP280_DIG_P4_LSB_REG    =    0x94  
BMP280_DIG_P4_MSB_REG    =    0x95  
BMP280_DIG_P5_LSB_REG    =    0x96  
BMP280_DIG_P5_MSB_REG    =    0x97  
BMP280_DIG_P6_LSB_REG    =    0x98  
BMP280_DIG_P6_MSB_REG    =    0x99  
BMP280_DIG_P7_LSB_REG    =    0x9A  
BMP280_DIG_P7_MSB_REG    =    0x9B  
BMP280_DIG_P8_LSB_REG    =    0x9C  
BMP280_DIG_P8_MSB_REG    =    0x9D  
BMP280_DIG_P9_LSB_REG    =    0x9E  
BMP280_DIG_P9_MSB_REG    =    0x9F 


class BMP180(object):
	def __init__(self, address=BMP280_I2C_ADDRESS):
		self._address = address
		self._bus = smbus.SMBus(1)
		# Load calibration values.
		if self._read_byte(BMP280_ID_REG) == BMP280_ID_Value:
			self._load_calibration()
			ctrlmeas = 0xFF #BMP280_T_MODE_1 << 5 | BMP280_P_MODE_1 << 2 | BMP280_SLEEP_MODE;  
			config = 0x14   #BMP280_T_SB1 << 5 | BMP280_FILTER_MODE_1 << 2;
			self._write_byte(BMP280_CTRL_MEAS_REG, ctrlmeas); 
			self._write_byte(BMP280_CONFIG_REG, config)
		else:
			print("Read BMP280 id error!\r\n")
	def _read_byte(self,cmd):
		return self._bus.read_byte_data(self._address,cmd)
	
	def _read_u16(self,cmd):
		LSB = self._bus.read_byte_data(self._address,cmd)
		MSB = self._bus.read_byte_data(self._address,cmd+1)
		return (MSB	<< 8) + LSB

	def _read_s16(self,cmd):
		result = self._read_u16(cmd)
		if result > 32767:result -= 65536
		return result

	def _write_byte(self,cmd,val):
		self._bus.write_byte_data(self._address,cmd,val)

	def _load_calibration(self):
		"load calibration"
		
		""" read the temperature calibration parameters """
		self.dig_T1 =self._read_u16(BMP280_DIG_T1_LSB_REG)
		self.dig_T2 =self._read_s16(BMP280_DIG_T2_LSB_REG)
		self.dig_T3 =self._read_s16(BMP280_DIG_T3_LSB_REG)
		""" read the pressure calibration parameters """
		self.dig_P1 =self._read_u16(BMP280_DIG_P1_LSB_REG)
		self.dig_P2 =self._read_s16(BMP280_DIG_P2_LSB_REG)
		self.dig_P3 =self._read_s16(BMP280_DIG_P3_LSB_REG)
		self.dig_P4 =self._read_s16(BMP280_DIG_P4_LSB_REG)
		self.dig_P5 =self._read_s16(BMP280_DIG_P5_LSB_REG)
		self.dig_P6 =self._read_s16(BMP280_DIG_P6_LSB_REG)
		self.dig_P7 =self._read_s16(BMP280_DIG_P7_LSB_REG)
		self.dig_P8 =self._read_s16(BMP280_DIG_P8_LSB_REG)
		self.dig_P9 =self._read_s16(BMP280_DIG_P9_LSB_REG)
		
		# print(self.dig_T1)
		# print(self.dig_T2)
		# print(self.dig_T3)
		# print(self.dig_P1)
		# print(self.dig_P2)
		# print(self.dig_P3)
		# print(self.dig_P4)
		# print(self.dig_P5)
		# print(self.dig_P6)
		# print(self.dig_P7)
		# print(self.dig_P8)
		# print(self.dig_P9)
		
	def compensate_temperature(self,adc_T):
		"""Returns temperature in DegC, double precision. Output value of "1.23"equals 51.23 DegC."""
		var1 = ((adc_T) / 16384.0 - (self.dig_T1) / 1024.0) * (self.dig_T2)
		var2 = (((adc_T) / 131072.0 - (self.dig_T1) / 8192.0)  * ((adc_T) / 131072.0  - (self.dig_T1) / 8192.0)) * (self.dig_T3)
		self.t_fine = var1 + var2
		temperature = (var1 + var2) / 5120.0
		return temperature
			
	def compensate_pressure(self,adc_P):
		"""Returns pressure in Pa as double. Output value of "6386.2"equals 96386.2 Pa = 963.862 hPa."""
		var1 = (self.t_fine / 2.0) - 64000.0
		var2 = var1 * var1 * (self.dig_P6) / 32768.0
		var2 = var2 + var1 * (self.dig_P5) * 2.0
		var2 = (var2 / 4.0) + ((self.dig_P4) * 65536.0) 
		var1 = ((self.dig_P3) * var1 * var1 / 524288.0  + (self.dig_P2) * var1) / 524288.0 
		var1 = (1.0 + var1 / 32768.0) * (self.dig_P1) 

		if var1 == 0.0: 
			return 0 # avoid exception caused by division by zero  

		pressure = 1048576.0 - adc_P
		pressure = (pressure - (var2 / 4096.0)) * 6250.0 / var1
		var1 = (self.dig_P9) * pressure * pressure / 2147483648.0
		var2 = pressure * (self.dig_P8) / 32768.0  
		pressure = pressure + (var1 + var2 + (self.dig_P7)) / 16.0

		return pressure; 

	def get_temperature_and_pressure(self):
		"""Returns pressure in Pa as double. Output value of "6386.2"equals 96386.2 Pa = 963.862 hPa."""
		xlsb = self._read_byte(BMP280_TEMP_XLSB_REG)
		lsb =  self._read_byte(BMP280_TEMP_LSB_REG)
		msb =  self._read_byte(BMP280_TEMP_MSB_REG)

		adc_T = (msb << 12) | (lsb << 4) | (xlsb >> 4)
		temperature = self.compensate_temperature(adc_T)

		xlsb = self._read_byte(BMP280_PRESS_XLSB_REG)
		lsb =  self._read_byte(BMP280_PRESS_LSB_REG) 
		msb =  self._read_byte(BMP280_PRESS_MSB_REG) 

		adc_P = (msb << 12) | (lsb << 4) | (xlsb >> 4)
		pressure = self.compensate_pressure(adc_P)
		return temperature,pressure
	
if __name__ == '__main__':

	import time
	
	print("BMP280 Test Program ...\n")
	
	bmp280 = BMP180()
				
	while True:
		time.sleep(1)
		temperature,pressure = bmp280.get_temperature_and_pressure()
		print(' Temperature = %.2f C Pressure = %.3f kPa'%(temperature,pressure/1000))

File: misc2/BMP280/wiringPi/bmp280.c
#include <wiringPi.h>
#include <wiringPiI2C.h>
#include <stdio.h>
#include <math.h>
#include "bmp280.h"

short dig_T2,dig_T3,dig_P2,dig_P3,dig_P4,dig_P5,dig_P6,dig_P7,dig_P8,dig_P9;
unsigned short dig_T1,dig_P1;
int t_fine;
int fd;

/******************* BMP280 Driver Interface *****************************/
char I2C_readByte(int reg)
{	
	return (char)wiringPiI2CReadReg8(fd,reg);
}

unsigned short I2C_readU16(int reg)
{
	int MSB,LSB;
	LSB = I2C_readByte(reg);
	MSB = I2C_readByte(reg + 1);
	int value = (MSB << 8) +LSB;
	return (unsigned short)value;
}

short I2C_readS16(int reg)
{
	int result;
	result = I2C_readU16(reg);
	if (result > 32767)result -= 65536;
	return (short)result;
}
void I2C_writeByte(int reg,int val)
{
	wiringPiI2CWriteReg8(fd,reg,val);
}

/******************* BMP280 operation functions *****************************/
void load_calibration()
{
	/* read the temperature calibration parameters */  
	dig_T1 = I2C_readU16(BMP280_DIG_T1_LSB_REG);  
	dig_T2 = I2C_readS16(BMP280_DIG_T2_LSB_REG);
	dig_T3 = I2C_readS16(BMP280_DIG_T3_LSB_REG);
	
	/* read the pressure calibration parameters */   
	dig_P1 = I2C_readU16(BMP280_DIG_P1_LSB_REG);
	dig_P2 = I2C_readS16(BMP280_DIG_P2_LSB_REG);
	dig_P3 = I2C_readS16(BMP280_DIG_P3_LSB_REG);
	dig_P4 = I2C_readS16(BMP280_DIG_P4_LSB_REG);
	dig_P5 = I2C_readS16(BMP280_DIG_P5_LSB_REG);
	dig_P6 = I2C_readS16(BMP280_DIG_P6_LSB_REG);
	dig_P7 = I2C_readS16(BMP280_DIG_P7_LSB_REG);
	dig_P8 = I2C_readS16(BMP280_DIG_P8_LSB_REG);
	dig_P9 = I2C_readS16(BMP280_DIG_P9_LSB_REG);

	printf("dig_T1 = %d\r\n",dig_T1);
	printf("dig_T2 = %d\r\n",dig_T2);
	printf("dig_T3 = %d\r\n",dig_T3);
	printf("dig_P1 = %d\r\n",dig_P1);
	printf("dig_P2 = %d\r\n",dig_P2);
	printf("dig_P3 = %d\r\n",dig_P3);
	printf("dig_P4 = %d\r\n",dig_P4);
	printf("dig_P5 = %d\r\n",dig_P5);
	printf("dig_P6 = %d\r\n",dig_P6);
	printf("dig_P7 = %d\r\n",dig_P7);
	printf("dig_P8 = %d\r\n",dig_P8);
	printf("dig_P9 = %d\r\n",dig_P9);
	
}

unsigned char BSP_BMP280_Init(void)
{
	unsigned char id,ctrlmeas, config; 
	
	id = I2C_readByte(BMP280_ID_REG);  
		
	if(id == BMP280_ID_Value)
	{
		load_calibration();
		
		ctrlmeas = 0xFF; //BMP280_T_MODE_1 << 5 | BMP280_P_MODE_1 << 2 | BMP280_SLEEP_MODE;  
		config = 0x14;   //BMP280_T_SB1 << 5 | BMP280_FILTER_MODE_1 << 2; 
		
		I2C_writeByte(BMP280_CTRL_MEAS_REG, ctrlmeas); 
		I2C_writeByte(BMP280_CONFIG_REG, config);  	
		
		return 0;
	}
	else
	{
		printf("Read BMP280 id error!\r\n"); 
		return 1;
	}
}

/* Returns temperature in DegC, double precision. Output value of "1.23"equals 51.23 DegC. */  
double compensate_temperature(int adc_T)  
{  
	double var1, var2, temperature;  
	var1 = (((double) adc_T) / 16384.0 - ((double) dig_T1) / 1024.0) * ((double) dig_T2); 
	var2 = ((((double) adc_T) / 131072.0 - ((double) dig_T1) / 8192.0)  * (((double) adc_T) / 131072.0  
					- ((double) dig_T1) / 8192.0)) * ((double) dig_T3);  
	t_fine = ( int) (var1 + var2);  
	temperature = (var1 + var2) / 5120.0;  
	
	return temperature;  
}  
  
  
/* Returns pressure in Pa as double. Output value of "6386.2"equals 96386.2 Pa = 963.862 hPa */  
double compensate_pressure(int adc_P)  
{  
	double var1, var2, pressure;  

	var1 = ((double)t_fine / 2.0) - 64000.0; 
	var2 = var1 * var1 * ((double) dig_P6) / 32768.0;  
	var2 = var2 + var1 * ((double) dig_P5) * 2.0;  
	var2 = (var2 / 4.0) + (((double) dig_P4) * 65536.0);  
	var1 = (((double) dig_P3) * var1 * var1 / 524288.0  + ((double) dig_P2) * var1) / 524288.0;  
	var1 = (1.0 + var1 / 32768.0) * ((double) dig_P1);  

	if (var1 == 0.0) {  
		return 0; // avoid exception caused by division by zero  
	}  

	pressure = 1048576.0 - (double) adc_P;  
	pressure = (pressure - (var2 / 4096.0)) * 6250.0 / var1;  
	var1 = ((double) dig_P9) * pressure * pressure / 2147483648.0;  
	var2 = pressure * ((double) dig_P8) / 32768.0;  
	pressure = pressure + (var1 + var2 + ((double) dig_P7)) / 16.0;  

	return pressure;  
}

/* Returns pressure in Pa as double. Output value of "6386.2"equals 96386.2 Pa = 963.862 hPa */  
void get_temperature_and_pressure(double *temperature, double *pressure)  
{  
	unsigned char lsb, msb, xlsb;  
	 int adc_P,adc_T;   

	xlsb = I2C_readByte(BMP280_TEMP_XLSB_REG);  
	lsb =  I2C_readByte(BMP280_TEMP_LSB_REG);  
	msb =  I2C_readByte(BMP280_TEMP_MSB_REG);  

	adc_T = (msb << 12) | (lsb << 4) | (xlsb >> 4); 
	//adc_T = 415148;
	* temperature = compensate_temperature(adc_T);

	xlsb = I2C_readByte(BMP280_PRESS_XLSB_REG);  
	lsb =  I2C_readByte(BMP280_PRESS_LSB_REG);  
	msb =  I2C_readByte(BMP280_PRESS_MSB_REG);  

	adc_P = (msb << 12) | (lsb << 4) | (xlsb >> 4); 
	//adc_P = 51988;
	* pressure = compensate_pressure(adc_P);     
} 

int main(int argc,char **argv)
{
	double temp,press;
	printf("BMP180 Test Program ...\n");
	if(wiringPiSetup() < 0) return 1;
	fd = wiringPiI2CSetup(BMP280_I2C_ADDRESS);
	load_calibration();
	while(1)
	{
		get_temperature_and_pressure(&temp,&press);
		printf("Temperature = %.2f C Pressure = %.3f kPa \r\n",temp,press/1000);
		delay(1000);
	}
	return 0;		
}

